### PR TITLE
feat(lib): Adds drain command to wash lib

### PR DESCRIPTION
**Pull request by thomastaylor312**
This also starts the process of creating a `config` module that I'll continue to update as I push forward the other PRs. Please note that this is the first of many PRs. I plan on doing each command as a separate PR instead of a mega PR

### CODE REVIEW

File: crates/wash-lib/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    Ok(wash)
+}
+
+pub fn model_cache_dir() -> IoResult<PathBuf> {
+    weld_codegen::weld_cache_dir().map_err(|e| Error::new(ErrorKind::Other, e.to_string()))
+}
+
+pub fn downloads_dir() -> IoResult<PathBuf> {
+    cfg_dir().map(|p| p.join(DOWNLOADS_DIR))
+}
```
Comment (brooksmtownsend) [inline crates/wash-lib/src/config.rs]: These should have a quick blurb about what these directories are used for

File: crates/wash-lib/src/drain.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
+use std::{env, fs, io::Result, path::PathBuf};
+
+use crate::config::{downloads_dir, model_cache_dir};
+
+/// A type that allows you to clean up (i.e. drain) a set of caches and folders used by wasmcloud
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "cli", derive(clap::Subcommand))]
```
Comment (brooksmtownsend) [inline crates/wash-lib/src/drain.rs:7]: niiiiice

File: crates/wash-lib/src/drain.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    /// Remove cached binaries extracted from provider archives
+    Lib,
+    /// Remove cached smithy files downloaded from model urls
+    Smithy,
+    /// Remove downloaded and generated files from launching wasmCloud hosts
+    Downloads,
+}
+
+// Drain::All.drain()
+// drain(DrainType::All)
```
Comment (brooksmtownsend) [inline crates/wash-lib/src/drain.rs]: I think these are leftover from the stream

File: crates/wash-lib/src/lib.rs [hunk 4]
```diff
@@ -8,3 +8,6 @@ pub mod start;
 
 #[cfg(feature = "parser")]
 pub mod parser;
+
+pub mod config;
+pub mod drain;
```
Comment (brooksmtownsend) [inline crates/wash-lib/src/lib.rs:13]: Thoughts on continuing the feature flag gating individual modules? I know that one per-module is probably overkill, but we could gate `drain` behind a flag. Config seems lightweight and applicable to most installation things

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]

File: crates/wash-lib/src/config.rs (hunk 1 above)
Comment (thomastaylor312) [reply crates/wash-lib/src/config.rs]: Oh whoops. Forgot the doc comments

File: crates/wash-lib/src/drain.rs (hunk 3 above)
Comment (thomastaylor312) [reply crates/wash-lib/src/drain.rs]: Yes, yes they are

File: crates/wash-lib/src/lib.rs (hunk 4 above)
Comment (thomastaylor312) [reply crates/wash-lib/src/lib.rs:13]: My plan was to add those gates at the _veeeerrrry_ end. It is a bit hard to tell where those gates should be until we get all done

Comment (brooksmtownsend) [reply crates/wash-lib/src/lib.rs:13]: Okay that sounds good to me, good call

**Review by brooksmtownsend — APPROVED** [review]

**Pull request merged**